new(async query): new NG doc (#8590)
newrelic/docs-website: src/content/docs/apis/nerdgraph/examples/async-queries-nrql-tutorial.mdx

---
title: "NerdGraph tutorial: Run NRQL queries asynchronously"
tags:
  - APIs
  - NerdGraph
  - Examples
metaDescription: Use New Relic NerdGraph to make asynchronous NRQL queries that have longer query duration limits.
---

> You can use our NerdGraph API to run an asynchronous NRQL query. This is useful for running queries that have less risk of being interrupted by browser or HTTP connection timeout limits or other issues. It's also especially applicable for running queries that may take a long time to complete.

Reviewer: ```suggestion
Our NerdGraph API can run an asynchronous NRQL query with `async: true`. If you have a query that may take a long time to complete, using an asynchronous NRQL query introduces less risk of query interruption by browser or HTTP connection timeout limits or other issues.
```

Reviewer: Curious if "other issues" is necessary -- I'd recommend either omitting it entirely unless it prevents interruption from any kind of issue. e.g., does this preserve the ~integrity of a query if there's infrastructure failure, or is only useful for interruptions? Not sure if that nuance makes sense but it might be worth teasing apart at least the genre of issues it reduces risk for

## Requirements [#requirements]

> The normal [NRQL query duration limit](/docs/query-your-data/nrql-new-relic-query-language/get-started/rate-limits-nrql-queries) is one minute. To run queries with a 10-minute limit, you must have [Data Plus](/docs/accounts/accounts-billing/new-relic-one-pricing-billing/data-ingest-billing#data-plus). (edited) 

Reviewer: Do you want (edited) to be here? Calling it out just in case

Author: ha. whoops

> Requires [Data Plus](/docs/accounts/accounts-billing/new-relic-one-pricing-billing/data-ingest-billing#data-plus). 

Reviewer 2: I'll defer to you on this, but to be clear: Data Plus isn't a requirement for asynchronous queries, it's just a requirement for query durations that extend all the way out to 10 minutes. To me, this is a little confusing, because it reads like I need Data Plus to run async queries at all. But I also realize from the title that this doc is mainly concerned with 10-minute queries, so maybe it's fine as is.

Author: ah i see, right, sorry

## Asynchronous query example [#example]

> Here's an example of an asynchronous query NerdGraph call. The `async: true` is what makes the query asynchronous. This query will return results if it completes within the default timeout of 5 seconds; otherwise it will return `queryProgress` data for use in follow-up queries to the `actor.account.nrqlQueryProgress` field. 

Reviewer: I'd add a sentence of exposition because when I got to the second example, I wasn't quite sure what exactly I was looking at until I reread it. Some ideas:

> An asynchronous query will return the results if it completes within the default timeout of 5 seconds. If an async query exceeds 5 seconds, then the query will return a `queryProgress` ID, which you can use to get results or the progress status of that query exceeding 5 seconds. 

It's a little redundant from a language-perspective, but I think being redundant works in the service of absolute clarity. A technical question I have is:

- Can you use the `queryProgress` ID to get partial results? i.e., if I had a stacked query to retrieve 2 discrete sets of data an datum A was retrieved before the code could get datum B, would I be able to use `queryProgress` to return A and get progress on B? Idk if that's just a science fiction example but it occurred to me~ 
- I'm not entirely clear what the `actor.account.nrqlQueryProgress` field is. Can you use `actor.account.nrqlQueryProgress` to do what I was asking about above ^?

> Here's an example of an asynchronous query NerdGraph call. The `async: true` is what makes the query asynchronous. This query will return results if it completes within the default timeout of 5 seconds; otherwise it will return queryProgress data for use in follow-up queries to the `actor.account.nrqlQueryProgress` field. 

Reviewer 2: Probably want to put `queryProgress` in backticks since it's the name of a field.

```
{
   actor {
      account(id: <var>YOUR_ACCOUNT_ID</var>) {
         nrql(query: <var>"SELECT * FROM Transaction"</var>, async: true) {
            results

>                queryProgress {

Reviewer 2: Minor formatting tweak: this line needs to be "de-dented" two spaces

               queryId
               completed
               retryAfter
               retryDeadline
               resultExpiration
            }
         }
      }
   }
}
```

## Getting the results or status of an async query [#example-query-results]

> When you run the previous query, it will return a query ID, which you can use to get results or progress status for that query. 

Reviewer: ```suggestion
If your query exceeds the default timeout, it will return a query ID. Take that ID and run a second query to retrieve the results or progress status for that query. 
```
